### ISSUE TITLE
`ServicePattern` pollutes the top-level namespace

### ISSUE THREAD

**Issue by ollietreend** [DFE-Digital/dfe-analytics#136]
The `ServicePattern` module is defined as a top-level module, rather than being namespaced within the `DfE::Analytics` module. It therefore pollutes the top-level namespace of any Ruby app which uses the dfe-analytics gem.

https://github.com/DFE-Digital/dfe-analytics/blob/3e078f9fdd0e890449f90e7cd2f151c392d570ac/lib/dfe/analytics/shared/service_pattern.rb#L1-L8

For example, this caused some strange behaviour over in https://github.com/DFE-Digital/itt-mentor-services/pull/521#discussion_r1578212067 where we had to manually require our own `ServicePattern` module directly in a spec file to override the one provided by dfe-analytics.

I think ideally this module should be namespaced inside `DfE::Analytics` to avoid this from happening – plus to be consistent with the other classes in this gem, which all seem to be namespaced according to their file path.

In other words, I think [lib/dfe/analytics/shared/service_pattern.rb](https://github.com/DFE-Digital/dfe-analytics/blob/3e078f9fdd0e890449f90e7cd2f151c392d570ac/lib/dfe/analytics/shared/service_pattern.rb) should be changed to:

```ruby
# a template for creating service objects
module DfE
  module Analytics
    module Shared
      module ServicePattern
        # ...
      end
    end
  end
end
```
